crates/solver/src/liquidity_collector.rs

```diff
@@ -43,3 +45,136 @@ impl LiquidityCollecting for LiquidityCollector {
         Ok(amms)
     }
 }
+
+/// A liquidity source which might not be initialised on creation. Instead
+/// initialisation gets retried in a background task over and over until it
+/// succeeds. Until the liquidity source has been initialised no liquidity will
+/// be provided.
+pub struct BackgroundInitLiquiditySource<L> {
+    liquidity_source: Arc<Mutex<Option<L>>>,
+}
+
+impl<L> BackgroundInitLiquiditySource<L> {
+    /// Creates a new liquidity source which might only be initialized at a
+    /// later point in time.
+    pub fn new<I, F>(label: &str, init: I, retry_init_timeout: Duration) -> Self
+    where
+        I: Fn() -> F + Send + Sync + 'static,
+        F: Future<Output = Result<L>> + Send,
+        L: LiquidityCollecting + 'static,
+    {
+        let liquidity_source: Arc<Mutex<Option<L>>> = Default::default();
+        let inner = liquidity_source.clone();
+        tokio::task::spawn(
+            async move {
```

Comment:
I wonder if we should setup a metric for alerts here - otherwise, we may be silently running without UniV3 or BalV2 liquidity without anyone knowing.

It sounds like an "unactionable" alert, but my argument is that "notifying the team in the `#general` channel" should be the corresponding action. Not sure how this fits into the new on-call plan, (maybe @fleupold can share his view on this as well).

Reply:
Not sure how much good it would do if #general knew about that. I think the only difference would be that colleagues **might** not raise any issues in #oncall because they knew that liquidity is bad at the moment.
However, since the retry logic basically fixes the liquidity as soon as possible people might not even notice.

In any case if we decide that we want to have metrics I would add them in a separate PR.

Reply:
I think we should alert if we are running without Balancer and v3 liquidity for an extended period of time (e.g. 1h). This would be actionable or at least noteworthy to me as it significantly reduces the quality of service

Reply:
Alright, will follow up with a PR for metrics.